File: jpa/src/main/java/com/jpa/configuration/RestTemplateResponseErrorHandler.java
package com.jpa.configuration;

import org.springframework.http.HttpStatus;
import org.springframework.http.client.ClientHttpResponse;
import org.springframework.web.client.ResponseErrorHandler;

import java.io.IOException;

public class RestTemplateResponseErrorHandler implements ResponseErrorHandler {
    @Override
    public boolean hasError(ClientHttpResponse response) throws IOException {
        HttpStatus.Series series = response.getStatusCode().series();
        return series == HttpStatus.Series.CLIENT_ERROR || series == HttpStatus.Series.SERVER_ERROR;
    }

    @Override
    public void handleError(ClientHttpResponse response) throws IOException {
        String msgError = "StatusCode: " + response.getStatusCode().value() + " "
                + response.getStatusCode().getReasonPhrase();
        if(response.getStatusCode().series() == HttpStatus.Series.SERVER_ERROR) {
            throw new IOException(msgError);
        }else if(response.getStatusCode().series() == HttpStatus.Series.CLIENT_ERROR) {
            throw new IOException(msgError);
        }
    }
}

File: security/src/main/kotlin/com/security/config/exceptionHandler/RestTemplateResponseErrorHandler.kt
package com.security.config.exceptionHandler

import org.springframework.http.HttpStatus
import org.springframework.http.client.ClientHttpResponse
import org.springframework.web.client.ResponseErrorHandler
import java.io.IOException

internal class RestTemplateResponseErrorHandler : ResponseErrorHandler {
    @Throws(IOException::class)
    override fun hasError(response: ClientHttpResponse): Boolean {
        val series = response.statusCode.series()
        return series == HttpStatus.Series.CLIENT_ERROR || series == HttpStatus.Series.SERVER_ERROR
    }

    @Throws(IOException::class)
    override fun handleError(response: ClientHttpResponse) {
        val msgError = ("""StatusCode: ${response.statusCode.value()} ${response.statusCode.reasonPhrase}""")
        if (response.statusCode.series() == HttpStatus.Series.SERVER_ERROR) {
            throw IOException(msgError)
        } else if (response.statusCode.series() == HttpStatus.Series.CLIENT_ERROR) {
            throw IOException(msgError)
        }
    }
}